### DISCUSSION TITLE
Requesting a specific OpenGL version

### DISCUSSION THREAD

**Discussion by LordIdra** [slint-ui/slint#3428] [Q&A] [answered]
Hi,

I've been looking into Slint for the last few days and love it so far. I've been looking at the opengl_texture and opengl_underlay examples to understand how to render with raw OpenGL (via glow) alongside Slint, but one issue is that the get_proc_address function I get from GraphicsAPI::NativeOpenGL loads a 3.2 version of OpenGL. Is there any way to request that the context be created with a higher version, so that more up-to-date OpenGL/GLSL functionality can be used? I have looked into the source code for the backends to try and understand if something like this is possible, but the code is quite complex and I'm still not entirely sure how they work, so hopefully this isn't a silly question or impossible ask :)

**Comment by tronical** [comment, marked as answer]
That's a very good question and I'm not entirely sure of the correct answer either ;). We request a context that supports GLES 2.0, using [with_context_api()](https://docs.rs/glutin/latest/glutin/context/struct.ContextAttributesBuilder.html#method.with_context_api). I think I've seen this sometimes resulting in contexts being created that are actually GLES 3.0 for example, but AFAIK that's fine as long as for example the renderer only uses GLES 2.0 features. I've had a situation once where the use of a vertex array object with a slint underlay clashed because the underlying renderer didn't use VAOs and expected the pristine 2.0 state, but saving/restoring it properly fixed it.

What functionality specifically are you looking for, btw?


More generally, I think these kind of specific needs might be best covered in a way where the application can compose the setup, by picking for example the winit backend (we'd like to make that public), along with a renderer, and then use renderer-specific API for more fine-grained graphics setup. For example for the FemtoVG renderer you could set up the GL context yourself and just provide the [OpenGLInterface](https://docs.rs/slint/latest/slint/platform/femtovg_renderer/trait.OpenGLInterface.html).

**Reply by LordIdra** [reply]
Thanks for the fast response. I'm not looking for any specific functionality per se, I'm just used to working with OpenGL version 3.3+, and seeing my #version 330 shaders fail to compile was a bit of a surprise. From my brief investigation it does like the approach with providing the OpenGLInterface myself would work, but also adds quite a bit of complexity... I think it's best I just work with a lower version of GL for now while I get used to Slint. If I do end up desperately needing something specific from a higher version later on I'll know roughly what to do though. Thanks :+1:

**Reply by tronical** [reply]
Hmm maybe there's a way though. Perhaps we can request a higher version and fall back to 2.0 if not supported. I think the renderer should be fine anyway with a newer version. Feel like experimenting a little? In glcontext.rs you could try bumping the minimum GLES version.
